File: server/src/main/kotlin/it/polito/server/security/saml/JwtAuthConverter.java
package it.polito.server.security.saml;

import org.springframework.core.convert.converter.Converter;
import org.springframework.security.authentication.AbstractAuthenticationToken;
import org.springframework.security.core.GrantedAuthority;
import org.springframework.security.core.authority.SimpleGrantedAuthority;
import org.springframework.security.oauth2.jwt.JwtClaimNames;
import org.springframework.security.oauth2.server.resource.authentication.JwtAuthenticationToken;
import org.springframework.stereotype.Component;
import org.springframework.security.oauth2.server.resource.authentication.JwtGrantedAuthoritiesConverter;
import org.springframework.security.oauth2.jwt.Jwt;

import java.util.Collection;
import java.util.Map;
import java.util.Set;
import java.util.stream.Collectors;
import java.util.stream.Stream;

@Component
public class JwtAuthConverter implements Converter<Jwt, AbstractAuthenticationToken> {
    private final JwtAuthConverterProperties properties;
    private final JwtGrantedAuthoritiesConverter jwtGrantedAuthoritiesConverter = new JwtGrantedAuthoritiesConverter();

    public JwtAuthConverter(JwtAuthConverterProperties properties) {
        this.properties = properties;
    }

    @Override
    public AbstractAuthenticationToken convert(Jwt jwt) {
        Collection<GrantedAuthority> authorities = Stream.concat(
                jwtGrantedAuthoritiesConverter.convert(jwt).stream(),
                extractResourceRoles(jwt).stream()
        ).collect(Collectors.toSet());
        return new JwtAuthenticationToken(jwt, authorities, getPrincipalClaimName(jwt));
    }

    private String getPrincipalClaimName(Jwt jwt) {
        String claimName = JwtClaimNames.SUB;
        if (properties.getPrincipalAttribute() != null) {
            claimName = properties.getPrincipalAttribute();
        }
        return jwt.getClaim(claimName);
    }

    private Collection<GrantedAuthority> extractResourceRoles(Jwt jwt) {
        try {
            @SuppressWarnings("unchecked")
            Map<String, Object> resourceAccess = jwt.getClaim("realm_access");
            @SuppressWarnings("unchecked")
            Collection<String> resourceRoles = (Collection<String>) resourceAccess.get("roles");
            return resourceRoles.stream().map(role -> new SimpleGrantedAuthority("ROLE_" + role))
                    .collect(Collectors.toSet());
        } catch (Exception e) {
            System.out.println("Error");
            return Set.of();
        }
    }

}

File: server/src/main/kotlin/it/polito/server/security/JwtAuthConverter.kt
package it.polito.server.security

import org.springframework.core.convert.converter.Converter
import org.springframework.security.authentication.AbstractAuthenticationToken
import org.springframework.security.core.GrantedAuthority
import org.springframework.security.core.authority.SimpleGrantedAuthority
import org.springframework.security.oauth2.jwt.Jwt
import org.springframework.security.oauth2.jwt.JwtClaimNames
import org.springframework.security.oauth2.server.resource.authentication.JwtAuthenticationToken
import org.springframework.security.oauth2.server.resource.authentication.JwtGrantedAuthoritiesConverter
import org.springframework.stereotype.Component
import java.util.stream.Collectors
import java.util.stream.Stream


@Component
class JwtAuthConverter(private val properties: JwtAuthConverterProperties) :
    Converter<Jwt?, AbstractAuthenticationToken?> {
    private val jwtGrantedAuthoritiesConverter = JwtGrantedAuthoritiesConverter()

    override fun convert(jwt: Jwt): AbstractAuthenticationToken {
        val authorities: MutableSet<GrantedAuthority>? = extractResourceRoles(jwt)?.stream()?.let {
            Stream.concat(
                jwtGrantedAuthoritiesConverter.convert(jwt)!!.stream(),
                it
            ).collect(Collectors.toSet())
        }

        return JwtAuthenticationToken(jwt, authorities, getPrincipalClaimName(jwt))
    }

    private fun getPrincipalClaimName(jwt: Jwt): String {
        var claimName: String? = JwtClaimNames.SUB
        if (properties.principalAttribute != null) {
            claimName = properties.principalAttribute
        }
        return jwt.getClaim(claimName)
    }


    private fun extractResourceRoles(jwt: Jwt): Set<SimpleGrantedAuthority>? {

        val resourceAccess = jwt.getClaim<Map<String, Any>>("resource_access")
        val resource: Map<String?, Any?>? = resourceAccess[properties.resourceId] as? Map<String?, Any?>
        val resourceRoles: Collection<String>? = resource?.get("roles") as? Collection<String>?
        return if (
            resourceAccess == null ||
            resourceAccess[properties.resourceId] == null ||
            resource?.get("roles") == null
            ) {
            setOf()
        } else {
            resourceRoles?.stream()?.map { role: String ->
                SimpleGrantedAuthority(
                    "ROLE_$role"
                )
            }?.collect(Collectors.toSet())
        }
    }
}
